test/js_test.go

```diff
@@ -1507,3 +1510,403 @@ func TestJetStreamSubscribe_AckDupInProgress(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		err = subB.Drain()
+		if err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+		fetchConsumers(t, 1)
+	})
+}
+
+func TestJetStream_UnsubscribeCloseDrain(t *testing.T) {
+	s := RunBasicJetStreamServer()
+	defer s.Shutdown()
+
+	if config := s.JetStreamConfig(); config != nil {
+		defer os.RemoveAll(config.StoreDir)
+	}
+
+	serverURL := s.ClientURL()
+	mc, err := nats.Connect(serverURL)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	jsm, err := mc.JetStream()
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	_, err = jsm.AddStream(&nats.StreamConfig{
+		Name:     "foo",
+		Subjects: []string{"foo.A", "foo.B", "foo.C"},
+	})
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	fetchConsumers := func(t *testing.T, expected int) []*nats.ConsumerInfo {
+		t.Helper()
+		cl := jsm.NewConsumerLister("foo")
+		if !cl.Next() {
+			if err := cl.Err(); err != nil {
+				t.Errorf("Unexpected error: %v", err)
+			}
+			t.Fatalf("Unexpected consumer lister next")
+		}
+		p := cl.Page()
+		if len(p) != expected {
+			t.Fatalf("Expected %d consumers, got: %d", expected, len(p))
+		}
+		if err := cl.Err(); err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+
+		return p
+	}
+
+	t.Run("conn drain deletes ephemeral consumers", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		nc, err := nats.Connect(serverURL, nats.ClosedHandler(func(_ *nats.Conn) {
+			cancel()
+		}))
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		js, err := nc.JetStream()
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		_, err = js.SubscribeSync("foo.C")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// sub.Drain() or nc.Drain() does not delete the durable consumers,
+		// just makes client go away.  Ephemerals will get deleted though.
+		nc.Drain()
+		<-ctx.Done()
+		fetchConsumers(t, 0)
+	})
+
+	jsm.Publish("foo.A", []byte("A.1"))
+	jsm.Publish("foo.B", []byte("B.1"))
+	jsm.Publish("foo.C", []byte("C.1"))
+
+	t.Run("conn close does not delete any consumer", func(t *testing.T) {
+		nc, err := nats.Connect(serverURL)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		js, err := nc.JetStream()
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		_, err = js.SubscribeSync("foo.A")
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		subB, err := js.SubscribeSync("foo.B", nats.Durable("B"))
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		resp, err := subB.NextMsg(2 * time.Second)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		got := string(resp.Data)
+		expected := "B.1"
+		if got != expected {
+			t.Errorf("Expected %v, got: %v", expected, got)
+		}
+		fetchConsumers(t, 2)
+
+		// There will be still all consumers since nc.Close
+		// does not delete ephemeral consumers.
+		nc.Close()
```

Comment:
Maybe we should be attempting to delete ephemeral consumers as well during `Close()`?

Reply:
We could, but server will clean them up on its on so not necessary.